todo_app/app.py

```diff
@@ -1,46 +1,104 @@
-from flask import Flask, render_template, redirect, url_for, request
+import os
+from flask import Flask, app, render_template, redirect, url_for, request, abort
+from functools import wraps
+import requests
+from flask_login import LoginManager, login_required, login_user, current_user
+from oauthlib.oauth2 import WebApplicationClient
 
 from todo_app.data.mongodb import MongoDB
+from todo_app.user import User
 from todo_app.view_model import ViewModel
 
+
 def create_app():
     app = Flask(__name__)
+    app.config['SECRET_KEY'] = os.getenv('SECRET_KEY')
+    app.config['LOGIN_DISABLED'] = os.getenv('LOGIN_DISABLED')
     database = MongoDB()
 
+    ## Auth Service Setup
+    login_manager = LoginManager()
+    login_manager.init_app(app)
+
+    client_id = os.getenv("AUTH_CLIENT_ID")
+    client_secret = os.getenv("AUTH_CLIENT_SECRET")
+    authorization_base_url = 'https://github.com/login/oauth/authorize'
+    token_url = 'https://github.com/login/oauth/access_token'
+
+    auth_client = WebApplicationClient(client_id)
+    auth_client.prepare_request_uri(authorization_base_url)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(auth_client.prepare_request_uri(authorization_base_url))
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
+
+    @app.route('/login/')
+    def login():
+        code = request.args.get('code')
+        token_request_url, token_request_headers, token_request_body = auth_client.prepare_token_request(token_url, authorization_response=request.url, client_secret=client_secret)
+        token_request_response = requests.post(token_request_url, headers=token_request_headers, data=token_request_body)
+        auth_client.parse_request_body_response(token_request_response.content.decode())
+        user_request_url, user_request_headers, user_request_body = auth_client.add_token("https://api.github.com/user")
+        user_request_response = requests.get(user_request_url, headers=user_request_headers, data=user_request_body)
+        user_name = user_request_response.json()['login']
+        user = User(user_name)
+        login_user(user)
+        return redirect(url_for('index'))
+
+    def writer_required(f):
+        @login_required
+        @wraps(f)
+        def decorated_function(*args, **kwargs):
+            user = User(current_user.id)
```

Comment:
Converting `current_user` to a `User` shouldn't be necessary. When you access the app, `load_user` sets the current_user object. And if you had to e.g. read the user from a database, you wouldn't want to have to do that in multiple places.

You may have found this necessary for your tests to pass however, because they have disabled login. What you actually want for your tests to set `login_manager.anonymous_user` to something.